### PR TITLE
feat: add pristine in branchCache 

### LINKED ISSUE #18464: Refactor parentSha/baseBranchSha logic in branch cache
### Describe the proposed change(s).

I think this might be a simpler solution:

- Remove parentSha altogether
- Add new boolean field "pristine" which is set to `true` every time a *new commit* is made (including branch creation) but set to `false` whenever base or branch SHA is changed by the user (or some commit event we didn't do)

Then for the purposes of merge confidence we will only include the data when `pristine=true`.

Cc @Gabriel-Ladzaretti

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
 - delete parentSha altogether
 - add a pristine (boolean) field instead which will be used for calculating mergeConfidence

## Context
 - Closes : #18464 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
This seems much more complex than I'd expected. I thought it would just be a boolean flag, for cache only, which is set to true on a new commit and set to false whenever baseBranchSha or branchSha change without one of our commits

File: lib/workers/repository/cache.ts [hunk 1]
```diff
@@ -52,12 +52,11 @@ async function generateBranchCache(
   try {
     const sha = getBranchCommit(branchName) ?? null;
     const baseBranchSha = getBranchCommit(baseBranch);
+    const pristine = getCachedPristineResult(branchName);
```
Comment (rarkins) [inline lib/workers/repository/cache.ts:55]: Why does it need to be read and then rewritten again?

Comment (RahulGautamSingh) [reply lib/workers/repository/cache.ts:55]: Maybe I didn't get you correctly? But if we don't do this how will we use the cached value on next run?

Comment (rarkins) [reply lib/workers/repository/cache.ts:55]: Shouldn't the value already be in the existing cache? Or do we "rewrite" the cache each run even if unchanged?

Comment (RahulGautamSingh) [reply lib/workers/repository/cache.ts:55]: We rewrite cache each run.

**Comment by rarkins** [comment]
Needs deconfliction

**Review by rarkins — DISMISSED** [review]

File: lib/workers/repository/finalise/repository-statistics.spec.ts [hunk 2]
```diff
@@ -120,27 +119,24 @@ describe('workers/repository/finalise/repository-statistics', () => {
...
+            branchName: 'dev',
+            sha,
+          },
+        ],
+        branches: [
+          { ...expectedMeta, branchName: 'b1' },
+          { ...expectedMeta, branchName: 'b2' },
+        ],
+        inactiveBranches: ['b3'],
+      });
```
Comment (Gabriel-Ladzaretti) [inline lib/workers/repository/finalise/repository-statistics.spec.ts:139]:
```suggestion
        ],
          inactiveBranches: ['b3'],
        },
        `Branch summary`
      );
```

Comment (RahulGautamSingh) [reply lib/workers/repository/finalise/repository-statistics.spec.ts]: Didnt work. It seems the cache is not getting initialized in the second test. Will need some debugging.

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**